flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PythonCalcMergeRule.java

```diff
@@ -0,0 +1,167 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.rules.logical;
+
+import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalCalc;
+import org.apache.flink.table.planner.plan.utils.PythonUtil;
+
+import org.apache.calcite.plan.RelOptRule;
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.core.Calc;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexFieldAccess;
+import org.apache.calcite.rex.RexInputRef;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexProgram;
+import org.apache.calcite.rex.RexProgramBuilder;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+/**
+ * Rule will merge Python {@link FlinkLogicalCalc} used in Map operation, Flatten {@link FlinkLogicalCalc}
+ * and Python {@link FlinkLogicalCalc} used in Map operation together.
+ */
+public class PythonCalcMergeRule extends RelOptRule {
+
+	public static final PythonCalcMergeRule INSTANCE = new PythonCalcMergeRule();
+
+	private PythonCalcMergeRule() {
+		super(operand(FlinkLogicalCalc.class,
+			operand(FlinkLogicalCalc.class,
+				operand(FlinkLogicalCalc.class, none()))),
+			"PythonCalcMergeRule");
+	}
+
+	@Override
+	public boolean matches(RelOptRuleCall call) {
+		FlinkLogicalCalc bottomCalc = call.rel(0);
+		FlinkLogicalCalc middleCalc = call.rel(1);
+		FlinkLogicalCalc topCalc = call.rel(2);
+		return isChainedRowBasedPythonFunction(bottomCalc, middleCalc, topCalc);
+	}
+
+	private boolean isChainedRowBasedPythonFunction(
+		FlinkLogicalCalc bottomCalc, FlinkLogicalCalc middleCalc, FlinkLogicalCalc topCalc) {
+		RexProgram bottomProgram = bottomCalc.getProgram();
+		List<RexNode> bottomProjects = bottomProgram.getProjectList()
+			.stream()
+			.map(bottomProgram::expandLocalRef)
+			.collect(Collectors.toList());
+
+		if (bottomProjects.size() != 1 || PythonUtil.isNonPythonCall(bottomProjects.get(0))) {
+			return false;
+		}
+
+		RexProgram topProgram = topCalc.getProgram();
+		List<RexNode> topProjects = topProgram.getProjectList()
+			.stream()
+			.map(topProgram::expandLocalRef)
+			.collect(Collectors.toList());
+		if (topProjects.size() != 1 || PythonUtil.isNonPythonCall(topProjects.get(0))) {
+			return false;
+		}
+
+		RexProgram middleProgram = middleCalc.getProgram();
+		List<RexNode> middleProjects = middleProgram.getProjectList()
+			.stream()
+			.map(middleProgram::expandLocalRef)
+			.collect(Collectors.toList());
+
+		return isInputsCorrespondWithUpstreamOutputs(bottomProjects, middleProjects) &&
```

Comment:
We should ensure the `condition` field of the `middleProgram ` is null here.

Reply:
Yes. Good catch.